### DISCUSSION TITLE
Doesn't load after update to 101.0.180 for WPF .NET 6.0

### DISCUSSION THREAD

**Discussion by pornganizer** [cefsharp/CefSharp#4118] [General]
- **What version of the product are you using?**
CefSharp.Wpf.NETCore -Version 101.0.180

- **What architecture x86 or x64?**
    x64
	
- **What version of .Net?**
![image](https://user-images.githubusercontent.com/21695807/169833458-3326f055-1fe0-4bb6-a41a-3b9829e4ae66.png)

    
- **On what operating system?**
    Win10

- **Are you using `WinForms`, `WPF` or `OffScreen`?**
    WPF
    
- **What steps will reproduce the problem?**
  Start the app and attempt to Initialize
```
            var settings = new CefSettings()
            {
                CachePath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "CefSharp\\Cache"),
                LogSeverity = LogSeverity.Default
            };

            Cef.Initialize(settings, performDependencyCheck: true, browserProcessHandler: null);
```

- **What is the expected output? What do you see instead?**
```
System.Exception: 'Unable to locate required Cef/CefSharp dependencies:
Missing:..\bin\Debug\net6.0-windows\runtimes\win-x64\lib\netcoreapp3.1\libcef.dll
Missing:..\bin\Debug\net6.0-windows\runtimes\win-x64\lib\netcoreapp3.1\icudtl.dat
Missing:..\bin\Debug\net6.0-windows\runtimes\win-x64\lib\netcoreapp3.1\snapshot_blob.bin
Missing:..\bin\Debug\net6.0-windows\runtimes\win-x64\lib\netcoreapp3.1\v8_context_snapshot.bin
Missing:..\bin\Debug\net6.0-windows\runtimes\win-x64\lib\netcoreapp3.1\resources.pak
Missing:..\bin\Debug\net6.0-windows\runtimes\win-x64\lib\netcoreapp3.1\chrome_100_percent.pak
Missing:..\bin\Debug\net6.0-windows\runtimes\win-x64\lib\netcoreapp3.1\chrome_200_percent.pak
Missing:..\bin\Debug\net6.0-windows\runtimes\win-x64\lib\netcoreapp3.1\libEGL.dll
Missing:..\bin\Debug\net6.0-windows\runtimes\win-x64\lib\netcoreapp3.1\libGLESv2.dll
Missing:..\bin\Debug\net6.0-windows\runtimes\win-x64\lib\netcoreapp3.1\d3dcompiler_47.dll
Missing:..\bin\Debug\net6.0-windows\runtimes\win-x64\lib\netcoreapp3.1\chrome_elf.dll
Missing:..\bin\Debug\net6.0-windows\runtimes\win-x64\lib\netcoreapp3.1\CefSharp.BrowserSubprocess.exe
Missing:..\bin\Debug\net6.0-windows\runtimes\win-x64\lib\netcoreapp3.1\CefSharp.BrowserSubprocess.Core.dll
Missing:..\bin\Debug\net6.0-windows\runtimes\win-x64\lib\netcoreapp3.1\icudtl.dat
Missing:..\bin\Debug\net6.0-windows\runtimes\win-x64\lib\netcoreapp3.1\libcef.dll
Missing:..\bin\Debug\net6.0-windows\runtimes\win-x64\lib\netcoreapp3.1\locales\en-US.pak
Executing Assembly Path:C..\bin\Debug\net6.0-windows\runtimes\win-x64\lib\netcoreapp3.1
```

- **Please provide any additional information below.**
  - No debug.log is generated
  - it worked before the update
  - the referenced directory contains only four files - i am not manipulating this directory in any way  
![image](https://user-images.githubusercontent.com/21695807/169832321-2c4a44d0-063c-4ec5-ba56-09a9c3e6b57a.png)

- **Does this problem also occur in the `CEF` Sample Application**
    - the sample application loads file

**Comment by amaitland** [comment]
Converting to discussion as not clear this is a bug. There haven't been any package changes for quite some time. 

Please test with https://github.com/cefsharp/CefSharp.MinimalExample

What version are you upgrading from?

Have you deleted your bin and obj folders? Or performed a clean checkout.

**Comment by pornganizer** [comment]
> Please test with https://github.com/cefsharp/CefSharp.MinimalExample

what do you want me to test? the wpf project loads and runs just fine, but the project is configured differently, note the samples target framework vs my wpf app
![image](https://user-images.githubusercontent.com/21695807/169933830-b2e22898-02ef-4d22-8d83-91b0f9594b2e.png)
![image](https://user-images.githubusercontent.com/21695807/169933840-fc8656de-7453-4b60-a44a-146e1f99d266.png)

> What version are you upgrading from?

nuget package version 100.0.230

> Have you deleted your bin and obj folders? Or performed a clean checkout.

yes i did read this as a common fix, tried it a bunch of times with VS closed even

**Reply by amaitland** [reply]
> what do you want me to test? the wpf project loads and runs just fine, but the project is configured differently, note the samples target framework vs my wpf app

How does your project differ? What does your csproj look like, not interested in the UI. 

Fork the minimal example, change the project so it matches yours, if the problem reproduces then push the changes to GitHub so I can compare.

> nuget package version 100.0.230

Only minor changes since then, you can easily compare https://github.com/cefsharp/CefSharp/compare/v100.0.230...v101.0.180

**Reply by amaitland** [reply]
Creating a new `.Net 6.0 WPF` application and installing `CefSharp.Wpf.NetCore` works as expected.

- Create new `WPF Appication` targeting `.Net 6.0`
- Install `CefSharp.Wpf.NETCore` version `101.0.180`
- Add `<cefsharp:ChromiumWebBrowser Address="http://google.com"/>` and xmlns
- Build
- Run

Application builds and starts as expected.

```xml
<TargetFramework>net6.0-windows</TargetFramework>
```

Adding a `RuntimeIdentifier` also works as expected (recommended).

```xml
<RuntimeIdentifier>win-x64</RuntimeIdentifier>
```

**Comment by pornganizer** [comment]
i am able to run a sample project with the same configuration as my project without any issues
i am not sure how to debug this further
i am being presented with an error message in `OnStartup` about cefsharp not finding files in a cefsharp created directory
i've been using cefsharp for over a year in this project without an issue until the latest nuget update

**Reply by amaitland** [reply]
No changes were made to the Nuget package structure, it's a simple minor upgrade so in general I don't expect any problems. 

- Perform a clean checkout from source control 
- Rollback changes and confirm previous version works as expected 
- Change msbuild output to normal or higher and look for errors/warnings in the log.
- Call CefSharpAfterBuildDiagnostic in your csproj and provide the output here https://github.com/cefsharp/CefSharp.MinimalExample/blob/cefsharp/101/CefSharp.MinimalExample.Wpf/CefSharp.MinimalExample.Wpf.netcore.csproj#L79

So far this is the only report of a problem, potentially the problem is specific to your environment.

**Reply by pornganizer** [reply]
> Rollback changes and confirm previous version works as expected

after downgrading to 100.0.230, my app no longer builds successfully

`Assembly 'CefSharp.Wpf' with identity 'CefSharp.Wpf, Version=101.0.180.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138' uses 'CefSharp.Core, Version=101.0.180.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138' which has a higher version than referenced assembly 'CefSharp.Core' with identity 'CefSharp.Core, Version=100.0.230.0`

do i need to do something other than `install-package cefsharp.wpf.netcore -version 100.0.230` to downgrade?

> Change msbuild output to normal or higher and look for errors/warnings in the log.

the log is quite large. is there anything i should be looking for specifically? here are some highlights containing cefsharp

```
warning NU1701: Package 'CefSharp.Wpf 101.0.180' was restored using '.NETFramework,Version=v4.6.1, .NETFramework,Version=v4.6.2, .NETFramework,Version=v4.7, .NETFramework,Version=v4.7.1, .NETFramework,Version=v4.7.2, .NETFramework,Version=v4.8' instead of the project target framework 'net6.0-windows7.0'. This package may not be fully compatible with your project.

Target "CefSharpPlatformCheck" skipped, due to false condition; (('@(AppConfigWithTargetPath)' == '' OR !Exists('$(CefSharpTransformXmlDllPath)')) AND '$(CefSharpPlatformTarget)' == 'AnyCPU' AND '$(CefSharpAnyCpuSupport)' != 'true' AND '$(CefSharpBuildAction)' != 'NoAction') was evaluated as (('' == '' OR !Exists('C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VisualStudio\v17.0\Web\Microsoft.Web.Publishing.Tasks.dll')) AND 'x64' == 'AnyCPU' AND 'true' != 'true' AND 'Content' != 'NoAction').

6>    Encountered conflict between 'Reference:C:\Users\...\.nuget\packages\cefsharp.wpf\101.0.180\lib\net452\CefSharp.Wpf.dll' and 'Reference:C:\Users\...\.nuget\packages\cefsharp.wpf.netcore\101.0.180\ref\netcoreapp3.1\CefSharp.Wpf.dll'. Could not determine winner due to equal file and assembly versions.
6>    Encountered conflict between 'CopyLocal:C:\Users\...\.nuget\packages\cefsharp.wpf\101.0.180\lib\net452\CefSharp.Wpf.dll' and 'CopyLocal:C:\Users\...\.nuget\packages\cefsharp.wpf.netcore\101.0.180\lib\netcoreapp3.1\CefSharp.Wpf.dll'. Choosing 'CopyLocal:C:\Users\...\.nuget\packages\cefsharp.wpf\101.0.180\lib\net452\CefSharp.Wpf.dll' arbitrarily as both items are copy-local and have equal file and assembly versions.
```

> Call CefSharpAfterBuildDiagnostic in your csproj and provide the output here

```
6>    CefSharp After Build Diagnostic
6>    CefSharpBuildAction = Default
6>    RuntimeIdentifier =
6>    CefSharpRuntimeIdentifier =
6>    Platform = AnyCPU
6>    PlatformName = AnyCPU
6>    Platforms = AnyCPU;x64
6>    PlatformTarget = x64
6>    PlatformTargetAsMSBuildArchitecture = x64
6>    TargetFramework = net6.0-windows
6>    TargetFrameworkVersion = v6.0
6>    NuGetProjectStyle = PackageReference
6>    SelfContained = false
6>    PublishSingleFile = true
6>    OutDir = bin\Debug\net6.0-windows\
6>    _CefSharpBrowserSubProcessFiles = C:\Users\...\.nuget\packages\cefsharp.common.netcore\101.0.180\runtimes\native\CefSharp.BrowserSubprocess.runtimeconfig.json
6>    IncludeAllContentForSelfExtract = true
6>    IncludeNativeLibrariesForSelfExtract =
```

> So far this is the only report of a problem, potentially the problem is specific to your environment.

most likely

**Reply by amaitland** [reply]
> do i need to do something other than `install-package cefsharp.wpf.netcore -version 100.0.230` to downgrade?

I'd suggest rolling back to a previous commit using source control. Your project file likely has mistakes.

> warning NU1701: Package 'CefSharp.Wpf 101.0.180' was restored using '.NETFramework,Version=v4.6.1, .NETFramework,Version=v4.6.2, .NETFramework,Version=v4.7, .NETFramework,Version=v4.7.1, .NETFramework,Version=v4.7.2, .NETFramework,Version=v4.8' instead of the project target framework 'net6.0-windows7.0'. This package may not be fully compatible with your project.

You appear to be referencing the `CefSharp.Wpf` package also somewhere in your build. This shouldn't be there if you are only targeting `.Net 6.0`

> PublishSingleFile = true

Remove `PublishSingleFile` from your csproj or set it to false.

> > So far this is the only report of a problem, potentially the problem is specific to your environment.
> 
> most likely

I'd say you have an incorrectly configured project file. Probably worked by good luck before.

**Reply by pornganizer** [reply]
> You appear to be referencing the CefSharp.Wpf package also somewhere in your build. This shouldn't be there if you are only targeting .Net 6.0

thank you so much!!
this was the key. my startup project was referencing a library that incorrectly had a reference to `CefSharp.Wpf` 
uninstalling it resolved the issue.

thanks so much!

**Comment by kpreisser** [comment]
Hi,
please try specifying `false` for `performDependencyCheck`:
```cs
Cef.Initialize(settings, performDependencyCheck: false, browserProcessHandler: null);
```
For .NET Core/.NET 5+, the native files (like `libcef.dll`) are in a different folder (`native` instead of `lib\netcoreapp3.1`) in the `runtimes` folder, which is why the dependency check won't find them when you don't publish the application using a platform-specific runtime identifier like `win-x64`.

**Reply by pornganizer** [reply]
If I specify false, the app crashes as soon as it tries to load a CefSharp control.

**Reply by kpreisser** [reply]
Sorry, you're right that this shouldn't be the problem, as in the current versions the `DependencyChecker` should search for the native files in the correct location for .NET Core/.NET 5+, so it should also work when `performDependencyCheck` is true.

**Reply by amaitland** [reply]
@kpreisser Are you experiencing any problems?

**Reply by kpreisser** [reply]
Hi @amaitland,
in normal usage I think the dependency check works; just when self-hosting the subprocess and not using a Runtime Identifier (with .NET Core/.NET 5+), the dependency check would fail as it would search for `libcef.dll` etc. in the directory of the app host (.exe), whereas the files would actually be in the `runtimes` subfolder in this case.

Therefore I first thought the dependency check might still generally fail for .NET Core/5+ without a RID, but then I realized from the code that the dependency check searches for the files in the browser subprocess directory (and the OP doesn't set a custom `BrowserSubprocessPath`), so it should work in this case.
Thanks!
